### PR TITLE
docker shim doc + kubelet flag update

### LINKED ISSUE #1440: Prepare for Docker-shim removal in K8s 1.24 - what is the impact for k0s?
### Is your feature request related to a problem? Please describe.

On next k8s release - 1.24 (expected to be released sometime in April) docker-shim will be completely removed (see https://kubernetes.io/blog/2022/01/07/kubernetes-is-moving-on-from-dockershim/)

### Describe the solution you would like

We should investigate the expected impact for k0s in k8s 1.24 and update documentation if necessary.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by trawler**
Signed-off-by: Karen Almog <kalmog@mirantis.com>

## Description

Fixes #1440

Documentation update on the impact of dockershim deprecation for kubernetes 1.24.
Kubelet flags are also updated to reflect the new change.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [ ] Auto test added

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: docs/runtime.md [hunk 1]
```diff
@@ -151,6 +151,10 @@ After editing the configuration, restart `k0s` to get containerd using the newly
 
 Use the option `--cri-socket` to run a k0s worker with a custom CRI runtime. the option takes input in the form of `<type>:<socket_path>` (for `type`, use `docker` for a pure Docker setup and `remote` for anything else).
 
-To run k0s with a pre-existing Docker setup, run the worker with `k0s worker --cri-socket docker:unix:///var/run/docker.sock <token>`.
+To run k0s with a pre-existing Docker setup, run the worker with `k0s worker --cri-socket docker:unix:///var/run/cri-dockerd.sock <token>`.
 
-When `docker` is used as a runtime, k0s configures kubelet to create the dockershim socket at `/var/run/dockershim.sock`.
\ No newline at end of file
+When `docker` is used as a runtime, k0s configures kubelet to create the dockershim socket at `/var/run/cri-dockerd.sock`.
```
Comment (jnummelin) [inline docs/runtime.md:156]: This is actually not true anymore as the dockershim is NOT managed by either k0s nor kubelet

File: docs/runtime.md [hunk 2]
```diff
@@ -151,6 +151,10 @@ After editing the configuration, restart `k0s` to get containerd using the newly
 
 Use the option `--cri-socket` to run a k0s worker with a custom CRI runtime. the option takes input in the form of `<type>:<socket_path>` (for `type`, use `docker` for a pure Docker setup and `remote` for anything else).
 
-To run k0s with a pre-existing Docker setup, run the worker with `k0s worker --cri-socket docker:unix:///var/run/docker.sock <token>`.
+To run k0s with a pre-existing Docker setup, run the worker with `k0s worker --cri-socket docker:unix:///var/run/cri-dockerd.sock <token>`.
```
Comment (jnummelin) [inline docs/runtime.md:154]: User also needs to setup the cri-dockerd shim, maybe just link to that page here?

**Review by jnummelin — CHANGES_REQUESTED** [review]

File: docs/docker-shim.md [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+```sh
+sudo systemctl daemon-reload
+sudo systemctl enable cri-docker.service
+sudo systemctl enable --now cri-docker.socket
+```
+
+### Configure K0s to use dockershim
+
+```sh
+sudo sed -i -e 's_^ExecStart.*_& --cri-socket docker:unix:///var/run/cri-dockerd.sock_' /etc/systemd/system/k0sworker.service
```
[the remark this answers is not in the document]
Comment (trawler) [reply docs/docker-shim.md:95]: No, it's an alternative separator for the sed command, so we don't have to escape every forward slash in the socket string.

File: docs/docker-shim.md [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+### Cordon and drain the node
+
+Get a list of all nodes:
+
+```sh
+sudo k0s kubectl get nodes -o wide
+
+NAME                                        STATUS   ROLES           AGE   VERSION       INTERNAL-IP   EXTERNAL-IP   OS-IMAGE             KERNEL-VERSION    CONTAINER-RUNTIME
+ip-10-0-49-188.eu-west-1.compute.internal   Ready    control-plane   3m    v1.24.0+k0s   10.0.49.188   <none>        Ubuntu 20.04.4 LTS   5.13.0-1022-aws   containerd://1.6.4
+ip-10-0-62-250.eu-west-1.compute.internal   Ready    <none>          31s   v1.24.0+k0s   10.0.62.250   <none>        Ubuntu 20.04.4 LTS   5.13.0-1017-aws   containerd://1.6.4
```
[the remark this answers is not in the document]
Comment (jnummelin) [reply docs/docker-shim.md:30]: hmm, but were "migrating" from docker to cri-dockerd+docker, right? So the _before_ state is still docker as the runtime, just via the built-in shim

[the remark this answers is not in the document]
Comment (trawler) [reply docs/docker-shim.md:30]: make sense... I'll update the doc

**Review by jnummelin — APPROVED** [review]

**Pull request merged**